tests/serve/utils/test_data_callback.py

```diff
@@ -47,7 +50,60 @@ def test_request_logging_mixed(tmp_dir):
         data_callback.log(request, response)
 
     assert os.path.exists(os.path.join(log_dir, "another.api.com-recordings.jsonl"))
-    assert os.path.exists(os.path.join(specs_dir, "another.api.com_mixed.yaml"))
+    assert os.path.exists(os.path.join(specs_dir, "another.api.com_mixed.json"))
+
+    with open(os.path.join(specs_dir, "another.api.com_mixed.json")) as f:
+        spec = convert_to_openapi(json.load(f))
+
+    assert "http://another.api.com" == spec.servers[0].url
+
+
+def test_request_logging_mixed_append(tmp_dir):
+    specs_dir = os.path.join(tmp_dir, "specs")
+    spec_path = os.path.join(specs_dir, "another.api.com_mixed.json")
+    os.makedirs(specs_dir)
+    with open(spec_path, "w") as f:
+        json.dump(convert_from_openapi(BASE_SCHEMA), f)
+
+    with open(spec_path) as f:
+        spec = convert_to_openapi(json.load(f))
+        assert spec.servers is None
+
+    request = RequestBuilder.from_dict(
+        dict(
+            method="get",
+            host="another.api.com",
+            pathname="/echo",
+            query={"message": "Hello"},
+            body="",
+            bodyAsJson={},
+            path="/echo",
+            protocol="http",
+            headers={},
+        )
+    )
+    response = ResponseBuilder.from_dict(
+        dict(
+            statusCode=200,
+            body='{"message": "hello"}',
+            bodyAsJson={"message": "hello"},
+            headers={},
+        )
+    )
+
+    log_dir = os.path.join(tmp_dir, "logs")
+    with RequestLoggingCallback(
+        log_dir=log_dir, specs_dir=specs_dir, update_mode=UpdateMode.MIXED
+    ) as data_callback:
+        data_callback.log(request, response)
+
+    assert os.path.exists(os.path.join(log_dir, "another.api.com-recordings.jsonl"))
+    assert os.path.exists(spec_path)
+
+    with open(spec_path) as f:
+        spec = convert_to_openapi(json.load(f))
```

Comment:
Maybe one could add a utility function like `read_openapi(filepath)` to avoid duplicating this both in tests and code base?

Reply:
I don't think 3 lines of code is a very bad copy-paste. Especially if one repetition is in tests. There's a more general function to read specs from a source. I don't think we need another simple one.

Reply:
With quick check I found eight different snippets in the code base for reading OpenAPI schemas from file system. I don't think that's very DRY :) But I can look how to improve on that next week.